tools/plugins/com.liferay.ide.project.core/src/com/liferay/ide/project/core/LiferayWorkspaceSDKListener.java

```diff
@@ -0,0 +1,123 @@
+import org.eclipse.core.resources.IResourceChangeEvent;
+import org.eclipse.core.resources.IResourceChangeListener;
+import org.eclipse.core.resources.IResourceDelta;
+import org.eclipse.core.resources.IResourceDeltaVisitor;
+import org.eclipse.core.resources.WorkspaceJob;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+
+/**
+ * @author Simon Jiang
+ */
+public class LiferayWorkspaceSDKListener implements IResourceChangeListener, IResourceDeltaVisitor
+{
+
+    private static final String ID_WORKSPACE_SDK_INVALID = "workspace-sdk-invalid";
+
+    @Override
+    public void resourceChanged( IResourceChangeEvent event )
+    {
+        if( event == null )
+        {
+            return;
+        }
+
+        try
+        {
+            event.getDelta().accept( this );
+        }
+        catch( CoreException e )
+        {
+        }
+    }
+
+    @Override
+    public boolean visit( final IResourceDelta delta ) throws CoreException
+    {
+        switch( delta.getResource().getType() )
+        {
+            case IResource.ROOT:
+            case IResource.FOLDER:
+            case IResource.PROJECT:
+                return true;
+
+            case IResource.FILE:
+            {
+                if ( delta.getKind() == IResourceDelta.ADDED || delta.getKind() == IResourceDelta.REMOVED )
+                {
+                    IFile deltaFile = (IFile)delta.getResource();
+
+                    if ( deltaFile.getName().equals( "build.properties" ))
+                    {
+                        Job job = new WorkspaceJob( "Checking workspace SDK" )
+                        {
+                            @Override
+                            public IStatus runInWorkspace( IProgressMonitor monitor ) throws CoreException
+                            {
+                                boolean findSDK = false;
+                                IProject[] projects = CoreUtil.getAllProjects();
+
+                                for( IProject existProject : projects )
+                                {
+                                    if ( existProject != null )
+                                    {
+                                        if ( SDKUtil.isValidSDKLocation( existProject.getLocation().toPortableString() ) )
+                                        {
+                                            ProjectUtil.clearMarkers( existProject, ID_WORKSPACE_SDK_INVALID,  IMarker.PROBLEM );
+
+                                            if ( findSDK == false )
+                                            {
+                                                findSDK = true;
+                                                existProject.refreshLocal( IResource.DEPTH_INFINITE, new NullProgressMonitor() );
```

Comment:
no need to call refreshLocal here if yo uare going to call it anyways outside of the if{} so this call or the other one isn't needed.